File: DotNet_Backend/SafarnamaApplication/SafarnamaApplication/Controllers/BookingController.cs
using Microsoft.AspNetCore.Mvc;
using SafarnamaApplication.DTOs;
using SafarnamaApplication.Services;

namespace SafarnamaApplication.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class BookingController : ControllerBase
    {
        private readonly IBookingService _bookingService;
        private readonly EmailServiceClient _emailServiceClient;
        private readonly CustomerService _customerService;

        public BookingController(IBookingService bookingService,    EmailServiceClient emailServiceClient, CustomerService customerService)
        {
            _bookingService = bookingService;
            _emailServiceClient = emailServiceClient;
            _customerService = customerService;
        }

        [HttpPost]
        public async Task<ActionResult<BookingDTO>> Create([FromBody] BookingDTO bookingDTO)
        {
            var created = await _bookingService.CreateBooking(bookingDTO);
            return Ok(created);
        }

        [HttpGet("{id}")]
        public async Task<ActionResult<BookingDTO>> Get(int id)
        {
            var booking = await _bookingService.GetBookingById(id);
            if (booking == null) return NotFound();
            return Ok(booking);
        }

        [HttpGet]
        public async Task<ActionResult<IEnumerable<BookingDTO>>> GetAll()
        {
            return Ok(await _bookingService.GetAllBookings());
        }

        [HttpPut("{id}")]
        public async Task<ActionResult<BookingDTO>> Update(int id, [FromBody] BookingDTO bookingDTO)
        {
            var updated = await _bookingService.UpdateBooking(id, bookingDTO);
            if (updated == null) return NotFound();
            return Ok(updated);
        }

        [HttpDelete("{id}")]
        public async Task<ActionResult> Delete(int id)
        {
            await _bookingService.DeleteBooking(id);
            return Ok("Booking deleted successfully");
        }

        [HttpGet("{id}/passengers")]
        public async Task<ActionResult<IEnumerable<PassengerDTO>>> GetPassengersForBooking(int id)
        {
            var dto = await _bookingService.GetBookingWithPassengers(id);
            if (dto == null || dto.Passengers == null) return NotFound();
            return Ok(dto.Passengers);
        }

        [HttpGet("test")]
        public ActionResult<string> Test()
        {
            return Ok("Controller is working");
        }

        //[HttpPut("{id}/payment")]
        //public async Task<ActionResult<string>> UpdatePaymentStatus(int id, [FromQuery] string status)
        //{
        //    var updated = await _bookingService.UpdatePaymentStatus(id, status);
        //    if (updated) return Ok("Payment status updated successfully");
        //    return NotFound();
        //}

        [HttpPut("{id}/payment")]
        public async Task<ActionResult<string>> UpdatePaymentStatus(int id, [FromQuery] string status)
        {
            var updated = await _bookingService.UpdatePaymentStatus(id, status);
            if (!updated) return NotFound();

            // 1. Get booking details
            var booking = await _bookingService.GetBookingById(id);
            if (booking == null) return NotFound();

            if (booking.CustomerId == null)
                return BadRequest("Booking does not have a customer linked.");

            // 2. Fetch customer details from CustomerService (or repository)
            var customer = await _customerService.GetCustomerByIdAsync(booking.CustomerId.Value);
            if (customer == null) return NotFound("Customer not found");

            // 3. Use customer info for the email
            var customerEmail = customer.CustEmail; // from CustomerDTO
            var subject = "Payment Confirmation";
            var body = $"Dear {customer.CustName}, your payment for booking #{booking.BookingId} was successful.";

            // 4. Call your Spring Boot email microservice
            await _emailServiceClient.SendEmailAsync(customerEmail, subject, body);

            return Ok("Payment status updated successfully, confirmation email sent to customer.");
        }


    }

}

File: Java_Backend/E-Tour/src/main/java/controller/BookingController.java
package controller;

import dto.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import service.BookingService;

import java.util.List;

@RestController
@RequestMapping("/api/bookings")
@CrossOrigin(origins="*")
public class BookingController {

    @Autowired
    private BookingService bookingService;

    @PostMapping
    public ResponseEntity<BookingDTO> create(@RequestBody BookingDTO bookingDTO) {
        return ResponseEntity.ok(bookingService.createBooking(bookingDTO));
    }

    @GetMapping("/{id}")
    
    public ResponseEntity<BookingDTO> get(@PathVariable Integer id) {
        BookingDTO dto = bookingService.getBookingById(id);
        return dto != null ? ResponseEntity.ok(dto) : ResponseEntity.notFound().build();
    }


    @GetMapping
    public ResponseEntity<List<BookingDTO>> getAll() {
        return ResponseEntity.ok(bookingService.getAllBookings());
    }

    @PutMapping("/{id}")
    public ResponseEntity<BookingDTO> update(@PathVariable Integer id, @RequestBody BookingDTO bookingDTO) {
        return ResponseEntity.ok(bookingService.updateBooking(id, bookingDTO));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<String> delete(@PathVariable Integer id) {
        bookingService.deleteBooking(id);
        return ResponseEntity.ok("Booking deleted successfully");
    }
    
    @GetMapping("/{id}/passengers")
    public ResponseEntity<List<PassengerDTO>> getPassengersForBooking(@PathVariable Integer id) {
        BookingDTO dto = bookingService.getBookingWithPassengers(id);
       
        if (dto == null || dto.getPassengers() == null) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(dto.getPassengers());
    }
    
    @GetMapping("/test")
    public ResponseEntity<String> test() {
        return ResponseEntity.ok("Controller is working");
    }

    @PutMapping("/{id}/payment")
    public ResponseEntity<String> updatePaymentStatus(@PathVariable Integer id, @RequestParam String status) {
        boolean updated = bookingService.updatePaymentStatus(id, status);
        if (updated) {
            return ResponseEntity.ok("Payment status updated successfully");
        } else {
            return ResponseEntity.notFound().build();
        }
    }
}
